src/zarr/indexing.py

```diff
@@ -2,84 +2,291 @@
 
 import itertools
 import math
+import numbers
+import operator
 from collections.abc import Iterator
-from typing import TYPE_CHECKING, NamedTuple
+from dataclasses import dataclass
+from enum import Enum
+from functools import reduce
+from types import EllipsisType
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    NamedTuple,
+    Protocol,
+    TypeGuard,
+    TypeVar,
+    cast,
+    runtime_checkable,
+)
 
-from zarr.common import ChunkCoords, Selection, SliceSelection, product
+import numpy as np
+import numpy.typing as npt
+
+from zarr.common import ChunkCoords, product
 
 if TYPE_CHECKING:
+    from zarr.array import Array
+    from zarr.buffer import NDArrayLike
     from zarr.chunk_grids import ChunkGrid
 
+BasicSelector = int | slice | EllipsisType
+BasicSelectorTuple = tuple[BasicSelector, ...]
+BasicSelection = BasicSelector | BasicSelectorTuple
+BasicSelectionNormalized = tuple[int | slice, ...]
+CoordinateSelection = npt.NDArray[np.intp]
+BlockSelector = int | slice
+BlockSelection = BlockSelector | tuple[BlockSelector, ...]
+BlockSelectionNormalized = tuple[BlockSelector, ...]
```

Comment:
I think when we use a type alias like `BlockSelectionNormalized` for `tuple[BlockSelector,...]` we lose some readability. To know that `BlockSelectionNormalized` is `tuple[int | slice, ...]` I have to do 2 lookups, and we aren't even saving lines of code because the type aliases are longer than the types. Was there a particular problem caused by the plain type aliases?